### ISSUE TITLE
PANIC: zfs: allocating allocated segment

### ISSUE THREAD

**Issue by samuelbernardo** [openzfs/zfs#6976]
### System information

Type                                | Version/Name
  ---                                  |     --- 
Distribution Name       | Gentoo
Distribution Version    | profile user customized
Linux Kernel                 | 4.9.69
Architecture                 | x86_64
ZFS Version                  | 0.7.3
SPL Version                  | 0.7.3

### Describe the problem you're observing

Kernel panic after starting OS with a force reboot and repete same operation that crashed the system before. I was doing a system upgrade and when calculating dependencies for the upgrade, kernel panic appeared.

After kernel panic system seems to be working normally without doing any action.
```
zpool status -x
all pools are healthy
```
```
zpool status -v
  pool: zfs
 state: ONLINE
status: Some supported features are not enabled on the pool. The pool can
        still be used, but some features are unavailable.
action: Enable all features using 'zpool upgrade'. Once this is done,
        the pool may no longer be accessible by software that does not support
        the features. See zpool-features(5) for details.
  scan: none requested
config:

        NAME                                                STATE     READ WRITE CKSUM
        zfs                                                 ONLINE       0     0     0
          raidz1-0                                          ONLINE       0     0     0
            ata-TOSHIBA_DT01ACA300_Z5RS6H0KS                ONLINE       0     0     0
            ata-TOSHIBA_HDWD130_678KR5JAS                   ONLINE       0     0     0
            ata-TOSHIBA_DT01ACA300_16QUEEEKS                ONLINE       0     0     0
            ata-TOSHIBA_HDWD130_678KPYLAS                   ONLINE       0     0     0
          raidz1-1                                          ONLINE       0     0     0
            ata-TOSHIBA_HDWD130_678KTDUAS                   ONLINE       0     0     0
            ata-TOSHIBA_DT01ACA300_16QUDE3KS                ONLINE       0     0     0
            ata-WDC_WD40EZRX-75SPEB0_WD-WCC4E2YAA98J-part6  ONLINE       0     0     0
            ata-TOSHIBA_HDWD130_678KPP7AS                   ONLINE       0     0     0
        cache
          sdc                                               ONLINE       0     0     0
          sde                                               ONLINE       0     0     0

errors: No known data errors
```

### Describe how to reproduce the problem

This is the configuration of zfs volume that had the bug (using deduplication and lz4 compression):

NAME   SIZE  ALLOC   FREE  EXPANDSZ   FRAG    CAP  DEDUP  HEALTH  ALTROOT
zfs  21.8T  5.69T  16.1T         -     6%    26%  1.07x  ONLINE  -
  raidz1  10.9T  2.85T  8.03T         -     6%    26%
    ata-TOSHIBA_DT01ACA300_Z5RS6H0KS      -      -      -         -      -      -
    ata-TOSHIBA_HDWD130_678KR5JAS      -      -      -         -      -      -
    ata-TOSHIBA_DT01ACA300_16QUEEEKS      -      -      -         -      -      -
    ata-TOSHIBA_HDWD130_678KPYLAS      -      -      -         -      -      -
  raidz1  10.9T  2.85T  8.03T         -     6%    26%
    ata-TOSHIBA_HDWD130_678KTDUAS      -      -      -         -      -      -
    ata-TOSHIBA_DT01ACA300_16QUDE3KS      -      -      -         -      -      -
    ata-WDC_WD40EZRX-75SPEB0_WD-WCC4E2YAA98J-part6      -      -      -         -      -      -
    ata-TOSHIBA_HDWD130_678KPP7AS      -      -      -         -      -      -
cache      -      -      -         -      -      -
  sdc   699G  79.7M   699G         -     0%     0%
  sde   699G  78.2M   699G         -     0%     0%

capacity  |   operations  |   bandwidth  |  total_wait   |  disk_wait  |  syncq_wait  |  asyncq_wait | scrub

pool |  alloc |  free |  read | write |  read | write |  read | write |  read | write  | read | write |  read | write |  wait 
  --- |   --- |   --- |   --- |  --- |   --- |  --- |   --- |  --- |   --- |  ---  |  --- |  --- |   --- |  --- |   --- 
zfs     |    5.69T | 16.1T  |   81 |   109 |  500K |  950K |   4us  |  1us  |  4us | 543ns | 187ns  |  2ns  |  1us |   1us | 723ns

### Include any warning/errors/backtraces from the system logs
```
[  335.068216] PANIC: zfs: allocating allocated segment(offset=10858229538816 size=8192)
[  335.068221] Showing stack for process 2019
[  335.068223] CPU: 4 PID: 2019 Comm: txg_sync Tainted: P        W  O    4.9.69-gentoox99 #1
[  335.068224] Hardware name: ASUS All Series/X99-S, BIOS 3402 08/18/2016
[  335.068226]  ffffc9002eff3a48 ffffffff81790587 0000000000000003 ffff880fdcae7800
[  335.068228]  ffffc9002eff3a58 ffffffffa0edf1bd ffffc9002eff3b78 ffffffffa0edf302
[  335.068230]  ffff880ff47eb0c0 6c6c61203a73667a 20676e697461636f 657461636f6c6c61
[  335.068232] Call Trace:
[  335.068238]  [<ffffffff81790587>] dump_stack+0x4d/0x66
[  335.068242]  [<ffffffffa0edf1bd>] spl_dumpstack+0x3d/0x40 [spl]
[  335.068245]  [<ffffffffa0edf302>] vcmn_err+0x52/0xf0 [spl]
[  335.068247]  [<ffffffff8122e27c>] ? __slab_free+0xbc/0x300
[  335.068249]  [<ffffffff8122d97e>] ? kmem_cache_alloc+0x12e/0x1c0
[  335.068251]  [<ffffffffa0eda8be>] ? spl_kmem_cache_alloc+0x5e/0xae0 [spl]
[  335.068252]  [<ffffffff8122e25c>] ? __slab_free+0x9c/0x300
[  335.068255]  [<ffffffffa0dc8616>] ? avl_insert+0xb6/0xd0 [zavl]
[  335.068273]  [<ffffffffa138bcba>] zfs_panic_recover+0x5a/0x60 [zfs]
[  335.068286]  [<ffffffffa13740d7>] range_tree_add+0x167/0x2a0 [zfs]
[  335.068298]  [<ffffffffa1373f70>] ? range_tree_destroy+0x60/0x60 [zfs]
[  335.068300]  [<ffffffff8122e67b>] ? kmem_cache_free+0x1bb/0x1e0
[  335.068311]  [<ffffffffa1373f70>] ? range_tree_destroy+0x60/0x60 [zfs]
[  335.068321]  [<ffffffffa1373f70>] ? range_tree_destroy+0x60/0x60 [zfs]
[  335.068331]  [<ffffffffa1374658>] range_tree_vacate+0x48/0xc0 [zfs]
[  335.068341]  [<ffffffffa1373f70>] ? range_tree_destroy+0x60/0x60 [zfs]
[  335.068354]  [<ffffffffa136ff04>] metaslab_sync_done+0x114/0x6c0 [zfs]
[  335.068369]  [<ffffffffa13938b9>] vdev_sync_done+0x39/0x70 [zfs]
[  335.068382]  [<ffffffffa137d7e3>] spa_sync+0x603/0xfd0 [zfs]
[  335.068384]  [<ffffffff81117d4d>] ? default_wake_function+0xd/0x10
[  335.068402]  [<ffffffffa139032c>] txg_rele_to_sync+0x73c/0x8d0 [zfs]
[  335.068415]  [<ffffffffa1390050>] ? txg_rele_to_sync+0x460/0x8d0 [zfs]
[  335.068417]  [<ffffffffa0edc260>] ? __thread_exit+0x20/0xa0 [spl]
[  335.068419]  [<ffffffffa0edc2cd>] __thread_exit+0x8d/0xa0 [spl]
[  335.068421]  [<ffffffff8110d232>] kthread+0xd2/0xf0
[  335.068423]  [<ffffffff8110d160>] ? kthread_park+0x60/0x60
[  335.068424]  [<ffffffff8110d160>] ? kthread_park+0x60/0x60
[  335.068427]  [<ffffffff81fdb112>] ret_from_fork+0x22/0x30
```

**Comment by samuelbernardo** [comment]
After repeating the update operation the load of the system start to increase indefinitely and when doing a reboot it only restart forcefully.
In the next reboot the mapping kernel panic appeared immediately after zfs import.

I reboot with live cd (ubuntu clonezilla) and load zfs module with zfs_recover=1. With `zpool import <pool name> -f` the kernel panic happened again.
The only solution I see now is mount read only, safeguard important data and creating the pool again.

I would ask you to review the available zfs tools so this kind of file system problems could be solved without the need to create the pool again.

**Comment by samuelbernardo** [comment]
I try another alternative zpool import before starting from scratch again:
`zpool import -fFN <pool name>`

but receive the kernel panic again:
VERIFY3(range_tree_space(rt) == space) failed (67477078016 == 67477086208)
PANIC at space_map.c:127:space_map_load()

Next step I'll destroy and create the pool again :(

**Comment by filip-paczynski** [comment]
Hi Samuel, do You have, by any chance `logbias=throughput` set on any dataset/zvol?

**Comment by samuelbernardo** [comment]
Hi filip-paczynski,

I always use the default value:

zfs   logbias               latency                default

I have already created the same pool again after wipefs all drives, and place all rescued data over the file system.

I upgraded to 0.7.4 with self made gentoo ebuild that I'll publish in my overlay. Let see if patches applied to 0.7.4 solve this bug.

**Comment by filip-paczynski** [comment]
I had a similar problem around 0.7.1-0.7.2 and it was related to `logbias=throughput`, seems that yours is different. Hopefully 0.7.4 will help.

**Comment by sempervictus** [comment]
I doubt patches would solve that one... Might be importable as read-only to get the data off of it though. This is _exactly_ the sort of thing ZFS is supposed to prevent though, convolution over whether or not data has been written successfully to a location on disk. It looks reminiscent of the original spacemap problems we found years ago on much older builds, but we did end up having to send/recv the data off to a new pool for all of them once the condition which caused those in the first place wasnt being repeated.

**Comment by samuelbernardo** [comment]
Good year!

I've updated to 0.7.5 but haven't yet reproduced the same actions with same environment.
Since I've been so unlucky with a ZFS with about 5TB of data with binary big files and many small files (over 1M), this time I only restored the critical data and leave in backup disks all less important data for daily work.

I'm also using btrfs as boot file system. I like zfs because it has better compression algorithms available, inline dedupe, better subvolume definition, provides zvols and have a better cache integration for storage devices allowing SSDs and HDDs to be placed where they are better. btrfs is far away from this... I believe that zfs module will be merged into linux kernel before btrfs could be comparable to zfs.

I would like to help zfs on linux project, but now I have little time since I'm finishing my thesis.

Thanks for your help to get zfs working on linux kernel.
Cheers

**Comment by krichter722** [comment]
I'm experiencing

```
[ 3059.763420] PANIC: zfs: allocating allocated segment(offset=1061113909248 size=4096) of (offset=1061113790464 size=135168)

[ 3059.763424] Showing stack for process 2301
[ 3059.763427] CPU: 5 PID: 2301 Comm: z_wr_iss Tainted: P           OE    4.16.0-041600rc4-generic #201803041930
[ 3059.763428] Hardware name: LENOVO 20221/INVALID, BIOS 71CN51WW(V1.21) 07/12/2013
[ 3059.763429] Call Trace:
[ 3059.763438]  dump_stack+0x63/0x8b
[ 3059.763449]  spl_dumpstack+0x42/0x50 [spl]
[ 3059.763453]  vcmn_err+0xc3/0x100 [spl]
[ 3059.763456]  ? spl_kmem_cache_alloc+0x73/0x7a0 [spl]
[ 3059.763461]  ? __slab_alloc+0x20/0x40
[ 3059.763465]  ? _cond_resched+0x19/0x40
[ 3059.763467]  ? kmem_cache_alloc+0x109/0x1c0
[ 3059.763529]  zfs_panic_recover+0x69/0x90 [zfs]
[ 3059.763534]  ? avl_find+0x5f/0x90 [zavl]
[ 3059.763568]  range_tree_add_impl+0x18b/0x460 [zfs]
[ 3059.763605]  ? dnode_rele+0x39/0x40 [zfs]
[ 3059.763644]  range_tree_add+0x11/0x20 [zfs]
[ 3059.763677]  space_map_load+0x2bc/0x4c0 [zfs]
[ 3059.763714]  metaslab_load+0x36/0xf0 [zfs]
[ 3059.763743]  metaslab_activate+0x93/0xc0 [zfs]
[ 3059.763745]  ? _cond_resched+0x19/0x40
[ 3059.763773]  metaslab_alloc+0x496/0x1140 [zfs]
[ 3059.763807]  zio_dva_allocate+0xac/0x630 [zfs]
[ 3059.763811]  ? spl_kmem_alloc+0xdf/0x1a0 [spl]
[ 3059.763813]  ? _cond_resched+0x19/0x40
[ 3059.763814]  ? mutex_lock+0x12/0x40
[ 3059.763818]  ? tsd_hash_search.isra.3+0x47/0xa0 [spl]
[ 3059.763822]  ? tsd_get_by_thread+0x2e/0x40 [spl]
[ 3059.763825]  ? taskq_member+0x18/0x30 [spl]
[ 3059.763858]  zio_execute+0x90/0xf0 [zfs]
[ 3059.763862]  taskq_thread+0x2ab/0x4e0 [spl]
[ 3059.763865]  ? wake_up_q+0x80/0x80
[ 3059.763896]  ? zio_reexecute+0x400/0x400 [zfs]
[ 3059.763899]  kthread+0x121/0x140
[ 3059.763903]  ? taskq_thread_should_stop+0x70/0x70 [spl]
[ 3059.763905]  ? kthread_create_worker_on_cpu+0x70/0x70
[ 3059.763907]  ret_from_fork+0x35/0x40
```

after heavy I/O operations while moving files from a btrfs onto a ZFS dataset with `dedup` enabled with

```
> sudo zpool status
  pool: data
 state: ONLINE
  scan: none requested
config:

	NAME        STATE     READ WRITE CKSUM
	data        ONLINE       0     0     0
	  sdb       ONLINE       0     0     0
	logs
	  sda1      ONLINE       0     0     0
	cache
	  sda3      ONLINE       0     0     0

errors: No known data errors
```

(one HDD with 2 SSD cache and logs partitions)

experienced with SPL 0.7.0-30_g3673d03 and ZFS 0.7.0-356_g639b18944 on Ubuntu 17.10 with Linux 4.16-rc mainline.

**Comment by kpande** [comment]
looks like some kind of spacemap corruption, does the system have ECC memory?

also, using shared non-mirrored device for log is really against the best practices, your slog latency is compromised by l2arc activity. you should use dedicated, mirrored devices.

**Comment by krichter722** [comment]
> looks like some kind of spacemap corruption, does the system have ECC memory?

@kpande The system has no ECC memory.

> also, using shared non-mirrored device for log is really against the best practices, your slog latency is compromised by l2arc activity. you should use dedicated, mirrored devices.

Unfortunately, I can't on my laptop. Does mirroring two partitions on the same SSD avoid/reduce the probability of spacemap corruptions (note you don't need to answer if you don't want to pollute the issue with support-scoped advice)?

**Comment by tlvu** [comment]
> The only solution I see now is mount read only, safeguard important data and create the pool again.

@samuelbernardo I have the same "PANIC: zfs: allocating allocated segment" error as you.  Now my machine lock up when I try to `zpool import -f <my pool>`.

How do you mount read-only to perform data recovery?  Did that even work?

**Comment by tlvu** [comment]
> How do you mount read-only to perform data recovery? Did that even work?

Replying to myself: `sudo zpool import -o readonly=on -f <pool name>`,  and the import works.  Starting data recovery now ... phew !!!
